java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -20,41 +20,105 @@
 package org.sonar.java.checks.unused;
 
 import com.google.common.collect.ImmutableList;
+
 import org.sonar.check.Rule;
 import org.sonar.java.RspecKey;
+import org.sonar.java.checks.helpers.MethodsHelper;
 import org.sonar.java.checks.serialization.SerializableContract;
 import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
 import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
 import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
 import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
 
+import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 @Rule(key = "UnusedPrivateMethod")
 @RspecKey("S1144")
 public class UnusedPrivateMethodCheck extends IssuableSubscriptionVisitor {
 
+  private final List<MethodTree> unusedPrivateMethods = new ArrayList<>();
+  private final Set<String> unresolvedMethodNames = new HashSet<>();
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR);
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    super.scanFile(context);
+    reportUnusedPrivateMethods();
+    unusedPrivateMethods.clear();
+    unresolvedMethodNames.clear();
+  }
+
+  private void reportUnusedPrivateMethods() {
+    unusedPrivateMethods.stream()
+      .filter(methodTree -> !unresolvedMethodNames.contains(methodTree.simpleName().name()))
+      .forEach(methodTree -> {
+        IdentifierTree simpleName = methodTree.simpleName();
+        reportIssue(simpleName, String.format("Remove this unused private \"%s\" %s.", simpleName.name(), methodTree.is(Tree.Kind.CONSTRUCTOR) ? "constructor" : "method"));
+      });
   }
 
   @Override
   public void visitNode(Tree tree) {
     if(!hasSemantic()) {
       return;
     }
-    MethodTree node = (MethodTree) tree;
-    Symbol symbol = node.symbol();
-    if (node.modifiers().annotations().isEmpty() && symbol.isPrivate() && symbol.usages().isEmpty()) {
-      if (node.is(Tree.Kind.CONSTRUCTOR)) {
-        if (!node.parameters().isEmpty()) {
-          reportIssue(node.simpleName(), "Remove this unused private \"" + node.simpleName().name() + "\" constructor.");
+    if (tree.is(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR)) {
+      checkIfUnused((MethodTree) tree);
+    } else {
```

Comment:
I would make explicit tests for every kind (or switch) and have overloaded method for those. That will make classCast less likely to happen. Here there is a lot of implicit going on. Let's make thing explicit.

checkIfUnknown(MethodInvocationTree ) 
checkIfUnknown(NewClasstree  )